components/raftstore/src/store/util.rs

```diff
@@ -795,11 +795,21 @@ pub enum ConfChangeKind {
 }
 
 impl ConfChangeKind {
-    pub fn confchange_kind(change_num: usize) -> ConfChangeKind {
-        match change_num {
+    pub fn confchange_kind(change_peers: &[ChangePeerRequest]) -> ConfChangeKind {
+        match change_peers.len() {
             0 => ConfChangeKind::LeaveJoint,
             1 => ConfChangeKind::Simple,
-            _ => ConfChangeKind::EnterJoint,
+            _ => {
+                if change_peers
+                    .iter()
+                    .find(|&cp| cp.get_change_type() != ConfChangeType::RemoveNode)
+                    .is_some()
+                {
+                    ConfChangeKind::EnterJoint
+                } else {
+                    ConfChangeKind::Simple
+                }
```

Comment:
Don't seem like a good idea to me... it could affect normal change peer command. 

How about adding a flag to `RaftRequestHeader` (see `WriteBatchFlags::TRANSFER_LEADER_PROPOSAL` for example) and work around the check based on this flag

Reply:
Added a flag to special-case the unsafe recovery conf change in `ChangePeerV2Request`, this is more accessible than adding it into the `RaftRequestHeader`, e.g. util.rs. PTAL